### ISSUE TITLE
storage: restarted node in need of snapshots can be wedged for long time

### ISSUE THREAD

**Issue by tbg** [cockroachdb/cockroach#37906]
repro steps:

- set up a three node tpcc-2k cluster (any warehouse count will do, but with more data the problem is more obvious)
- run tpcc with --tolerate-errors and --wait=false
- kill one node, but leave everything else running
- wait a few minutes (maybe 10? more if it doesn't repro)

Restart the dead node and run a SELECT COUNT(*) on the tpcc tables through it

Chances are this will hang for minutes before returning. The reason is that none of the replicas will have been moved off that node while it was down (since there are only three nodes in the cluster), but now all of them will need Raft snapshots. There is no particular ordering in which these snapshots are requested. Some of our COUNT(*) requests (or even worse, liveness requests etc) may hit a replica on the local node in need of a snapshot. The lease request will hang until the snapshot is applied, which can take a long time.

We should do something here, like short-circuit lease request (for a blank NotLeaseholderErr) if we can with reasonable accuracy detect that the node is in need of a snapshot. This isn't trivial to detect on a follower (would be easier on the leader, but alas). We could also prioritize snapshots for system ranges higher to resolve the somewhat orthogonal problem that such hung requests on the system range can prevent the node from becoming live in the first place, which causes hard-to-diagnose problems.

**Comment by andy-kimball** [comment]
Can the leader detect if one or more of the Raft followers is behind, and if so, delay Raft log truncation until the dead node timeout expires (i.e. 5 minutes)?

**Comment by tbg** [comment]
Raft log truncation at the time of writing takes into account node liveness in the sense that it will immediately become willing to "sacrifice" followers that are not live (i.e. after ~4-5s of downtime). As we discussed in the weekly, this amounts to a historical artifact (because we had to be very careful to not let the Raft log grow large historically due to its inclusion in snapshots) that isn't a concern any more as apparent from this issue the current behavior is undesirable.

I suggest the following heuristics:

- if the Raft log is excessively large (say > 2x larger than a snapshot would be, and at least a few MB - maybe much more; tuning this is TBD), sacrifice the down followers (i.e. truncate the log). Also do this if the store is low on disk. Otherwise,
- if there is any down follower, don't truncate (regardless of how long the node has been down).

The second step may seem unexpectedly simple, but I think it checks out. If we sacrifice followers the moment they're marked as dead they will still have most of their replicas (removing them requires sending snapshots elsewhere, so it'll go at a leisurely pace compared to how fast the log queue can truncate). A dead node's replicas are supposedly going to be transferred away completely over time OR the node will restart, at which point log truncation will become possible again (since no dead follower exists any more). And the first criterion protects us from amassing a prohibitively long Raft log. At the same time, it decouples the log queue from the store dead timeout, which is a nice property.

**Comment by tbg** [comment]
The other problem we have here is that replicas in need of a snapshot translate the latency of queueing for and finally receiving the snapshot to any clients waiting on the replica. If we can detect the situation in which a replica is waiting for a snapshot without a timeout, we're in a good place to prevent this from becoming a problem because we can refuse the requests with a `NotLeaseholderError` which should prompt DistSender to try another replica (eventually discovering and caching the true leaseholder). Ideally we can achieve this without digging into the Raft internals too much, though it's even unclear what the Raft internals provide. On a follower, we don't even receive the latest commit index, and we're also not really sure if we will receive a snapshot or not. But there will be some way we can reliably figure this out

However, one spicy problem is that when a request first arrives at such a range, the result will be a lease request proposed to Raft. This lease request's outcome will only be known after the snapshot arrives (and the replica can catch up all the way to the result of the lease request). As a result, *if* the replica actually succeeds in acquiring the lease (it may well if no other lease has been active since the replica went down), no other replica can serve traffic until the stuck replica receives its snapshot, effectively stalling that range for that duration (which can be many hours depending on the amount of data that needs to be transferred). We could "probe" Raft (with a timeout) before issuing any lease requests and could build a small state machine around the probes that can translate into a circuit breaker to solve the first problem above. "Probing" could either mean submitting a noop through Raft, or sending a one-off RPC to a quorum of replicas and noting the highest committed index, which would be required to be surpassed by that of the local replica before declaring the probe a success.

I think to defuse this issue to 80% we need to fix the log truncation and come up with a relatively reliable indicator for returning a NotLeaseholderError, falling back to a timeout on the lease request to prevent the worst case of requests getting stuck indefinitely.

The situation in which the down node gets a lease is hopefully rare because it needing a snapshot sort of implies that there was another leader in the meantime that wrote more log entries.

**Comment by bdarnell** [comment]
Could the quota pool cause this to escalate into even more severe failures? If the node is up and responding to heartbeats but can't accept new log entries until it gets a snapshot, it will prevent the release of proposal quota and then block writes even on the live nodes.

**Comment by tbg** [comment]
@bdarnell the quota pool handles this case, though we'll need to check that it actually works as advertised, I don't think it ever mattered before.

https://github.com/cockroachdb/cockroach/blob/c097a16427f65e9070991f062716d222ea5903fe/pkg/storage/replica_proposal_quota.go#L204-L209

**Comment by danhhz** [comment]
I can reproduce this readily with a ~10m roachtest that's basically Tobi's original instructions. In addition to absurdly long response times on SQL queries, I sometimes can't even get a SQL connection to the just restarted node (getting user info?). (I'm also curious if this effect is the underlying cause of what I was seeing in #36879.)

I've looked at this today and I think we need two mitigations. First, when a node is down for a bit (~minutes), we make it less likely for it to need a snapshot when it is restarted. Second, we lessen the impact if it does need snapshots. I've focused first on things that are hopefully safe enough to backport into 19.1.3, to unblock our affected customer. 19.1.3 is cut a week from Monday, so I'm hoping to get fixes in soon so they can bake for a week and shake out any unforeseen consequences.

To avoid snapshots when a node comes back, we'll make the log truncation less aggressive. I think something like what's described in https://github.com/cockroachdb/cockroach/issues/37906#issuecomment-505919802 is a good long term solution, but to me it feels too risky to backport without more baking time than we have. A good compromise is to simply raise the threshold for what the raft log queue considers "recently active". (The raft log queue doesn't truncate when a follower is missing but has been recently active.) Recent currently has a threshold of 10s and shares the same MaxQuotaReplicaLivenessDuration constant as the quota pool. This sharing seems to have been introduced in #32439 but not for any particular principled reason. A natural place to raise the threshold to is the `server.time_until_store_dead` cluster setting, which is the time we give up on the node and start transferring away its replicas.

Experimentally this seems to work decently well. There are still a small number of snapshots, but it's small enough that the node recovers quickly when it comes back up. We should investigate why there are still this small number of snapshots, but in the meantime the second mitigation (lessening the impact of shapshots on a node that's just come up) will cover the rest. Note that this only helps if the node is down for less than `server.time_until_store_dead`, which defaults to 5m. If I leave the node down for 7m, there is a rush of raft log truncations at 5m and when it comes back up, it needs tons of snapshots. However, our customer can temporarily raise the `server.time_until_store_dead` setting while doing the rolling migration and lower it when the migration is done.

For the second mitigation, I think we need to have some sort of timeout on redirectOnOrAcquireLease, will will return NotLeaseholderError when it fires. redirectOnOrAcquireLease currently blocks on getting the snapshot, which can be hours if every replica is getting one and this replica is at the end of the line. Relying on this timeout will have an affect on tail latencies, but so far, it's the least risky thing I've thought of. Some more sophisticated ideas have been tossed around to do this without the tail latency, but it's nice to have a low-risk, obvious-how-it-works CYA here. I have yet to run experiments for this. The two big open questions are what threshold to use for the timeout and what happens if the timeout fires on the leaseholder?

**Comment by tbg** [comment]
I posted some discussion of the log truncation approach on the review in https://github.com/cockroachdb/cockroach/pull/38484. The TL;DR is that the most intuitive safe code change seems to not care about the 5 minute distinction, which is nice. We leave nodes alone when they're up, and when they're not up we also leave them alone unless the log is large, in which case we sacrifice them. (Today the "leave them alone unless" isn't there, we go straight to sacrificing by pretending the followers aren't even there).

For the timeout in `redirectOnOrAcquireLease`, I think we'll want to start with something conservative (i.e. in the at least a few seconds range. 5?). This is owed just to the fact that it'll go into a patch release and while a shorter timeout would solve the problem "better" it may also "do things" in unexpected places.

For implementing this, do you really have to do anything more than change this code

https://github.com/cockroachdb/cockroach/blob/c097a16427f65e9070991f062716d222ea5903fe/pkg/storage/replica_range_lease.go#L1028-L1032

to early return with the `NotLeaseholderError`, plus lowering duration until that branch fires?

**Comment by bdarnell** [comment]
A long-term solution to the redirectOnOrAcquireLease problem would be to replace raft-level proposal forwarding with an RPC that could return the index of the new proposal. Then we could know to stop waiting if our index is way behind the index of our proposed lease. (or we could forward the proposal with a maxIndex so it could be dropped if the leader is ahead). Replacing the proposal flow is also contemplated in #21849.

**Comment by tbg** [comment]
I proposed something similar in https://github.com/cockroachdb/cockroach/issues/37906#issuecomment-505925527, but I like the idea of making this about Raft proposal forwarding. If the follower includes its commit index into the proposal the leader will be in a good position to make a reasonable call via heuristics, though this necessarily involves some magic values which would be good to avoid. Instead, the leader could refuse proposals unless the follower is in `StateReplicate`. There are lots of details to figure out with this, but it seems promising.

**Comment by tbg** [comment]
A first mitigation has landed on master and is very likely to be included in the next 19.1.x patch release. While this first mitigation has solved the problem in our testing and should greatly reduce it in most real-world scenarios too, we are working on additional safeguards related to this issue:

- prevent lease requests from hanging
- prevent leadership to be attained by node that is far behind.

**Comment by bdarnell** [comment]
Should we backport #38484 to 2.1 as well?

**Comment by danhhz** [comment]
Does 2.1 have the work that removes the raft log from snapshots? If not, we might want to do 2.1-specific stability testing of that PR before backporting.

**Comment by tbg** [comment]
No, that's 19.1. With 2.1 we run a higher risk of seeing that snapshot refusal loop again (I think snapshots with 8mb are refused, and truncation would kick in at 4mb). I believe the refusal code was often triggered by SSTs which clock in at ~22mb each from what I remember, so changing the truncation heuristics likely wouldn't make a big difference for the worse (but it should make one for the better from what we can tell). But we'd need to do our due diligence there, I'm thinking that it would pass the new roachtest 5x and a reasonably sized import (1+ hour) that passes reliably (are there any of them?) before the cherry pick should continue to do so after.

**Comment by RoachietheSupportRoach** [comment]
Zendesk ticket #[3433](https://cockroachdb.zendesk.com/agent/tickets/3433) has been linked to this issue.

**Comment by roncrdb** [comment]
@tbg 

Had a chat with a user about this issue today.  While they are pleased that their downtime is no longer 2 hours (now just down to minutes).  They are still concerned about the fact that there is still some downtime where the nodes go into an "unready" state when a node is restarted.

**Comment by nvanbenschoten** [comment]
> As a result, if the replica actually succeeds in acquiring the lease (it may well if no other lease has been active since the replica went down), no other replica can serve traffic until the stuck replica receives its snapshot, effectively stalling that range for that duration

I think it's very unlikely that an epoch-based lease would ever get in the situation where its lease request actually succeeds because it needs to propose the lease acquisition with a lease sequence (see `RaftCommand.ProposerLeaseSequence`). If this lease sequence isn't correct then the corresponding Raft command will be rejected beneath Raft in `checkForcedErr`. However, if the lease sequence is correct then why was the restarted follower trying to acquire a new one? For expiration-based leases, it might just not have observed a recent extension, but for epoch-based leases, this is all handled through node liveness.

> Instead, the leader could refuse proposals unless the follower is in StateReplicate. There are lots of details to figure out with this, but it seems promising.

@danhhz and I just talked about this issue and it looks like we came to pretty much the same idea. Our thought was to have the leader reject forwarded MsgApps from followers who are in StateSnapshot and propagate the rejection back to the follower. We'd probably only want to do this for forwarded lease acquisition proposals to avoid interacting with things like replicas unquiescing (although this might be fine because a follower can't be in StateSnapshot while the range is quiesced). The rejection would then be passed back to the lease request as a NotLeaseholderError.

One concern here is that the forwarded MsgApp might be a reproposal of a previously proposed lease acquisition that could eventually succeed. Rejecting it eagerly along the proposal forwarding path might not be quite correct in that case. To get around this, we could be even more selective and only reject forwarded proposals that we can determine have no chance of succeeding (either in earlier or later reproposals) due to their `ProposerLeaseSequence`. This might be fairly expensive to check on the proposal forwarding path, but that's far from the hot path.

**Comment by tbg** [comment]
Since you just talked about this, what happened to the idea of sending a noop write through raft? The "leader refuses certain proposals" strategy seems a lot more involved and harder to reason about.

**Comment by danhhz** [comment]
Mostly I couldn't explain it well enough to get the idea across. I'm also curious what nathan's thoughts on this are

**Comment by tbg** [comment]
Related -- a snapshot isn't necessary to create a problem here. If a follower is behind by "two minutes worth of raft log entries" the same problem exists, but the leader will never have the follower as needing a snapshot.

@nvanbenschoten the noop write idea is basically:

Before sending a lease request, send a noop write through raft (it needs to be one that actually replicates all the way through, and can be proposed by any replica). If that doesn't work out within - say - 500ms, return NotLeaseholderError. If it does, go ahead and request the actual lease. This is like a "pre-vote" for trying to become the leaseholder.

**Comment by nvanbenschoten** [comment]
> If a follower is behind by "two minutes worth of raft log entries" the same problem exists, but the leader will never have the follower as needing a snapshot.

In theory, that's true, though in practice I thought the issue was that only a single snapshot can be ingested by a node at a time, so the problem compounds if more than one Range is hit by this because they all line up to be snapshotted.

> This is like a "pre-vote" for trying to become the leaseholder.

Ha, I think the word "pre-vote" was used in our discussion. My main concern with that idea is that it seems difficult to pick the 500ms threshold. We'd want to set it low enough that we didn't stall traffic unnecessarily but high enough to avoid starvation or false negatives, which could bubble all the way back to the client (DistSender doesn't retry forever, so we only have one shot per Replica). I don't have any major reservations about the idea though.

**Comment by danhhz** [comment]
If I wanted to try out the no-op write, what would I do? I was thinking maybe an Increment to a new range-local replicated key for the command. What's the write place to plug that in, evalAndPropose?

**Comment by tbg** [comment]
> In theory, that's true, though in practice I thought the issue was that only a single snapshot can be ingested by a node at a time, so the problem compounds if more than one Range is hit by this because they all line up to be snapshotted.

Yes, I'd expect this to be worse with snapshots, just pointing out that you don't need-need them. I'm still corresponding with the customer that is still reporting problems here. Their symptoms don't quite match my expectations, so it's not even clear that this is the problem they're experiencing.

> My main concern with that idea is that it seems difficult to pick the 500ms threshold.

Yeah, having to pick a threshold is awkward, but if you can't commit a command within, say, 1s, should you be requesting a lease? I wouldn't call this solution perfect, but it's simple and good, so I prefer it fairly strongly over relying on the Raft internals even more. I don't think our timing here has to be extremely aggressive. For a "perfect" solution we'd probably be best off sending an RPC to the other replicas to make the pre-voting more explicit. But then you're picking constants again

> DistSender doesn't retry forever, so we only have one shot per Replica

Are you sure? Exhausting the replicas results in a SendError, which retries indefinitely here:

https://github.com/cockroachdb/cockroach/blob/5119819560e40d5b6e4661d4dac08c8e9b6ebf4b/pkg/kv/dist_sender.go#L1461-L1476

@danhhz I think this is a bit more complicated than I thought earlier because of the need to "get the command into Raft". If it's just a regular old write, it'll get stuck trying to acquire the lease first, which of course we don't want. The only commands that avoid this are lease requests themselves, and so we need to add another exception here,

https://github.com/cockroachdb/cockroach/blob/0013508c54a7233ec9286eb94ed931b293d190e6/pkg/storage/replica_proposal.go#L712-L715

at which point it's natural to make a dedicated `NoopWriteRequest` (name tbd). If we wanted to avoid making a new request (for example to allow backporting) the next best thing is to do something like this

https://github.com/cockroachdb/cockroach/blob/dd5053ee464daab2b8f89f871f97469fc5fcb0a3/pkg/storage/split_delay_helper.go#L46-L57

which just puts an empty entry straight into raft, but then it won't be easy at all to determine when that entry is committed -- for that you need to make a `proposal`, and constructing a proposal is most straightforward when you hand the replica a request, at which point it's back to making a `NoopWriteRequest`.

A new request like this is also generally useful because we can use it to ascertain the liveness of a range.

If all you want is prototype, we can take shortcuts, but afaict we don't have a smoke test for the result, or do we?

**Comment by danhhz** [comment]
@tbg Sorry if i missed this, but what's your thought on doing this by handing a BatchRequest with an IncrementRequest directly to evalAndPropose?

> If all you want is prototype, we can take shortcuts, but afaict we don't have a smoke test for the result, or do we?

I started writing a unit test yesterday. It currently repros a call to `(Replica).Send` getting hung when the replica needs a snapshot. I haven't tried to repro the stuck replica actually succeeding in getting the lease because I'm not sure how to actually make that happen (for the reasons nathan described).

**Comment by bdarnell** [comment]
> Our thought was to have the leader reject forwarded MsgApps from followers who are in StateSnapshot and propagate the rejection back to the follower.

I think I like this approach better than the pre-write. There's some plumbing and backwards-compatibility concerns to introduce the response path, but it would let us solve #21849 at the same time. 

> The "leader refuses certain proposals" strategy seems a lot more involved and harder to reason about.

It doesn't seem so bad to me; the main thing I see is to make sure that we unquiesce the range and maybe resend a snapshot if one isn't in flight. 

> One concern here is that the forwarded MsgApp might be a reproposal of a previously proposed lease acquisition that could eventually succeed. 

Lease requests don't get reproposed (and per #21849, with the proper plumbing we'd substantially reduce reproposals of all kinds). And even if a lease request was falsely rejected (even though it seems unlikely that you'd fall back to needing a snapshot while a valid lease request is in flight), at worst you'd just bounce around a few times and come back to the right lease holder.

**Comment by tbg** [comment]
I agree that we should solve #21849 but it feels like the dog wagging the tail. Explaining that proposed mechanism is going to be difficult. So we rely on the fact that in that problematic situation where we're proposing the lease but are not behind, we're not the raft leader. So far, so good. We will try to get the lease and a proposal will get forward to the leader. The leader rejects it... why? The leader doesn't necessarily know that the follower needs a snapshot yet; it's reasonably likely that it's in probing status. Do we only allow forwarded proposals from peers in StateReplicate? There have been multiple bugs with this state tracking in the past and while at this point I trust that it mostly works, I dread every added time we rely on it. So anyway, let's say we figure out that we need to drop the proposal, the leader somehow has to connect that back to the client waiting on the follower. I guess we'd do this by instead proposing lease request that will be rejected under the same command ID? How else are we going to plumb the error back to the caller?

Also I'm probably missing something but I don't see how this ties in with https://github.com/cockroachdb/cockroach/issues/21849. Are you suggesting we disable proposal forwarding in general? I assume you're not because that means having to do the routing via an RPC of our own (may be fine, but certainly not in scope here). And I don't see where in the proposal here you'd ever see an `ErrProposalDropped`.

Anyway, I hope it's becoming clear that this is... not easy at all when you look from up closer. Between the two options I'm seeing one that's

- easy to understand and moderately easy to write, is going to solve the problem, also helps out in other ways (assess range health, handle the non-snap behind scenario)
- hard to understand and write, is going to solve the problem maybe a tiny bit better but also is going to eat at least a few bugs, also helps out in other ways (I don't see that yet, but I assume you'll tell me how).

Maybe I'm not being fair here (I tried anyway) but until there's a pro-con list that tilts in favor of not doing the noop write that's the one I think is the most appropriate.

**Comment by bdarnell** [comment]
> Also I'm probably missing something but I don't see how this ties in with #21849. Are you suggesting we disable proposal forwarding in general? I assume you're not because that means having to do the routing via an RPC of our own (may be fine, but certainly not in scope here). And I don't see where in the proposal here you'd ever see an ErrProposalDropped.

Yes, I was suggesting getting rid of raft-level proposal forwarding completely, replacing it with our own request/response rpc, and using that to solve both this and #21849 (aiming for 20.1, of course).

But I have another idea now:

> The leader rejects it... why?

Because the leader should try to get the lease instead. It can propose its own lease as well as it can forward the follower's. If it forwards the follower's lease that will lead to another disruption of availability as leadership is transferred. (maybe we'll transfer lease and leadership later, but I think it's probably better to leave that out of the lease bootstrapping process so we can use the better-controlled transfer process).

I think that raft followers forwarding RequestLease commands to raft leaders is a relic from before we had TransferLease. Now we can have a simple rule: Only raft leaders can attempt to acquire a lease on an unleased range. If a follower activates and sees there is a raft leader but no valid lease, it wakes up the leader instead of trying to get the lease itself (if there is no raft leader, it campaigns for leadership and then if it wins it tries to get the lease).

**Comment by danhhz** [comment]
@nvanbenschoten I'm not going to get to this before my OOO, can you handle the routing of whoever is going to take it over, please?

Latest status I've heard was that Tobi wasn't sure the leaseholder timeout is actually the issue that the customer was having, but I can't find where that discussion was happening. He may have just been looking at what's already been posted to the zendesk ticket.

I've got a branch with a test for the leaseholder timeout, I'll try to clean it up and PR it today.

**Comment by tbg** [comment]
I had asked for more info on Zendesk, and asked Ron to update Dan and I as
info comes in.

**Comment by ajwerner** [comment]
We saw this go on for a very very long time recently because the behind, restarted node needed a snapshot of a range which was over 500MB. At the default 8MB/s, sending this snapshot from the raftSnapshotQueue would always fail because that queue has a timeout of 1m. This is straightforward to work around by increasing the rate limit but is a bummer and took a long time to realize. 

In light of this and https://github.com/cockroachdb/cockroach/issues/39717 (increasing the default range size), I'm going to give the raftSnapshotQueue a much longer default timeout and am going to set an inner timeout based on the size of the snapshot being sent and the current rate limit.

**Comment by ajwerner** [comment]
@lunevalex pinging you on this rather severe issue as @drewdeally has hit it again. IIRC this was discussed as a high priority fix for 20.2, just want to give it some visibility so we can be explicit one way or the other.

**Comment by lunevalex** [comment]
@ajwerner thanks for looping me in, will look into this and figure out a course of action.

**Comment by andreimatei** [comment]
FWIW, I think this issue is the biggest thing to improve towards better a better rolling restart experience. I think it's big on @knz's agenda.

**Comment by knz** [comment]
We have a customer  possibly encountering this. 

What would be good `vmodule` settings to see this particular issue unfolding? I don't see leaseholder lookup failures and re-elections logged by default.

**Comment by ajwerner** [comment]
I've been thinking about this a bit. It seems like part of the reason we see this is that when nodes start up and get traffic, they first resolve the range descriptor. Then, given the total lack of entries in their leaseholder cache, they send to the closest range, which happens to be localhost. Would we be able to substantially mitigate this problem if we avoiding sending requests to the local node when the leaseholder cache is empty? I'm not saying this would be a solution but I suspect that it would go a very long way. 

@andreimatei you've touched the leaseholder cache and replica decision-making in the `DistSender` recently. Do you have thoughts?

**Comment by ajwerner** [comment]
I rescind this thought. For clusters with large numbers of ranges, there will still be leaseholders cached as the newly revived range. Perhaps if we were to clear the leaseholder cache entries corresponding to a node when that epoch ends or make the leaseholder cache epoch aware we could make ourselves more robust to this problem.

**Comment by andreimatei** [comment]
Why did you rescind this your thoughts, exactly?
The range cache is completely disconnected from the local replicas. And the cache is wiped on restart. So, I guess all requests are sent to the local replica where, if the old lease is not expired (which implies that it's owned by some other replica, not this one), a NLHE should be returned and the request is redirected. Which is good.
But if the old lease appears to be expired, then I think what happens is not so good. The local replica I guess will try to acquire a lease, and that stalls.

Discussions in this issue have been focused on mitigating the problem at the "server" level by not trying to acquire that lease. But perhaps they could also be mitigated at the client level by peeking into the local replica's lease before sending to it.

**Comment by ajwerner** [comment]
I rescinded the comment because I tried it out. I set up a `bank` workload on 4 nodes with `--rows=1000000000` and ran the workload over all of the nodes with `--tolerate-errors`. I think killed one of the nodes and brought it back. At first I thought I'd mitigate the problem because the `DistSender` on that local node wouldn't send anything to itself and then maybe we wouldn't get hung. What actually happened is that the workload never re-connected to that node and everything still got hung just due to stale leaseholder cache entries on the other nodes sending to that node.

It does seem like if we made the leaseholder cache epoch-aware we could mitigate this problem on the client side. The client should have no trouble figuring out if a cached leaseholder corresponds to an old epoch, but if it does discover that, what then? I'm rather unsure we could eliminate the problem with client-side mitigations and thus I'm not sure it's worth the effort of putting them in place.

**Comment by ajwerner** [comment]
A different client-side mitigation came up in conversation today. In cases where a node is being taken down for temporary maintainence that might take on the order of 10s of minutes, it might be reasonable to not drain the node of all of its data. This bug makes that less reasonable. Perhaps a way to unlock that decision would be to mark the current node as draining and then ensure that batch requests are not sent to this node if it is draining or if the epoch is mismatched.

**Comment by knz** [comment]
Discussed this with @tbg and @andreimatei yesterday. We were focusing on the point that Tobias identified as next action earlier last year:

> prevent lease requests from hanging

In short, a node starts and its replicas are "dormant". Upon the first request to each replica, it tries to acquire a lease, which stalls (because in  need of snapshot as discussed earlier). This is what we want to mitigate.

Here is what we discussed:

-  We expect most of the trouble to come from quiesced ranges. 

   If a range is currently active somewhere else in the cluster, the restarting node is likely to receive an update about the LH via the heartbeat mechanism before receiving arequest, and then all is well. (I write "likely" here because it's not so unlikely that a request is received before that heartbeat.)  

   Because of quiesced ranges, we can't just sit there and do nothing until we receive an LH heartbeat for all replicas.

- A strawman solution is to start out each replica as follower, and call out an election (send a raft pre-vote) upon receiving the first request if the replica is still dormant, to discover the leader.

  - We do it only upon receiving the first request, and not preemptively as soon as the node starts up, so as to avoid an election storm and also give a chance to natural cluster activity to preemptively inform us.

  - The way this works is that when there's no valid lease, before acquiring/proposing a lease we peek at raft. If we find the raft group is dormant, that's when we have the special case. The special case is to send a pre-vote. Then we need to "wait" somehow to discover the result.  If we don't discover anything, we send a NotLeaseHolderError back.

- We discussed that "wait until we discover the result" is difficult. Raft does not provide us a mechanism to synchronously wait on a response. We considered the following alternatives:

  - Trigger the pre-vote and then immediately send NotLeaseHolderError back to the client (distsender). In this solution we expect the client to retry indefinitely. Eventually the voting completes and the request will succeed. It's a sort of busy wait.

  - Trigger the pre-vote, then wait 1s unconditionally, then peek if we have a result already. If we do, then all is well and we can use the lease. If Raft hasn't completed the election yet, we send NotLEaseHolderError and fall back to the first solution above.

  - Trigger the pre-vote, then iteratively poll raft locally until we have a response.

- Tobias would like to avoid the wait altogether. Unclear how to achieve that yet (Tobias please provide your thoughts if you had them already).

**Comment by tbg** [comment]
> If a range is currently active somewhere else in the cluster, the restarting node is likely to receive an update about the LH via the heartbeat mechanism before receiving arequest, and then all is well. (I write "likely" here because it's not so unlikely that a request is received before that heartbeat.)

This isn't quite right. We were discussing when a Raft group would know its leader, not the leaseholder. And the statement is, then: if the range is not quiesced when the node comes back, the leader (not leaseholder) will reach out after at most one heartbeat interval and at that point the Raft instance on the slow follower would learn who the leader is.
The leaseholder becomes known by catching up on the log.

But the main point in your comment remains valid: since we have to assume that most ranges in an average cluster are quiesced, and since the node restarts with its raft groups without any information on past leaders, we have to find a solution that works sufficiently well even if all (many) replicas are in this state.
So there are two steps: first, we need to trigger discovery/election of raft leader. And we make it an invariant that we'll only request a lease if the leadership is local - otherwise we redirect to the leader instead. I'm not sure how lease constraints factor in here, though. We generally have the raft leader follow the lease, but here we're proposing the opposite. Maybe it's acceptable that the lease will start out "somewhere" and then promptly transfer to the right location? This is the status quo after all, so I hope this is fine.

I think the "wait until we discover the result" step is what will put us in the same trouble we are trying to solve here. The waiting will block a client request! We don't want to optimistically assume that this replica will get the lease, I think we need to be pessimistic and assume that if you don't know where the raft leader is, then it's better to go try someone else. So I favor the

> Trigger the pre-vote and then immediately send NotLeaseHolderError back to the client

option. This will work well (enough) whenever there is actually a replica that has the lease as we will discover it that way. If all replicas are on recently restarted nodes, they will all have campaigns triggered, and once a leader is elected (there could be a split election here which would incur another timeout, unfortunately, though in the common case I would expect the first replica to be touched to win, if it has the latest log entries, since it gets a good head start) the lease will be acquired. This is also what I referred to as the "avoid the wait" option, in the sense that it avoids the wait in the common case in which only one node is behind and the other replicas are fully initialized - we would bounce off the slow replica once, but the next attempt would either hit or discover the leaseholder.

Finally, I should add that in the first ~10 minutes (scanner interval) we wake up all of a recently restarted node's raft group once, so after that time they should all know where the leader is (or at least was at some point) and are properly quiesced in that state.

**Comment by tbg** [comment]
BTW all of this basically picks up @bdarnell's suggestion above: https://github.com/cockroachdb/cockroach/issues/37906#issuecomment-529041377

Except that we're currently not discussing a dedicated RPC from the slow follower to the leader but use NotLeaseholderError in its place. I like Ben's vision of disabling Raft proposal forwarding altogether, but I think there will be too many snags here (what if the leaseholder loses the raft leadership? Someone else will get it and then what?) and what we're proposing might be a step in the right direction, but without too drastic a change to existing behavior. I would be excited to see this prototyped.

**Comment by nvanbenschoten** [comment]
I like where this discussion is going but have a few questions with the recent proposal.

> I think that raft followers forwarding RequestLease commands to raft leaders is a relic from before we had TransferLease. Now we can have a simple rule: Only raft leaders can attempt to acquire a lease on an unleased range. If a follower activates and sees there is a raft leader but no valid lease, it wakes up the leader instead of trying to get the lease itself (if there is no raft leader, it campaigns for leadership and then if it wins it tries to get the lease).

I'm curious where this new rule is going to live. It seems most natural to me that it would live below Raft and would be encoded as a policy that RequestLease proposals cannot be forwarded and instead result in a failed proposal (with a NotLeaseholderError pointed at the leader we were planning to forward to?). This would allow the proposal of a RequestLease to trigger the leadership election transparently and only if necessary without us needing to muck around with Raft state and a manual pre-vote round above Raft. It would also keep the logic fully within the Raft abstraction, which is a big design goal we've been pushing towards.

The "avoid the wait" option then seems like an optimization on top of this, and one that isn't obviously worth the added complexity. It's worth noting that there are benefits to waiting synchronously while attempting to acquire the lease on the follower. Notably, if the NotLeaseholderError generated during the failed proposal forwarding pointed at the leader then we'd avoid blindly jumping around to random other followers. So the cost would be a single synchronous round of pre-vote to discover the current leader – that doesn't seem too bad and places an upper bound on how long this whole process takes, given stable leadership _somewhere_. Without it, is there an upper bound? I think it would be a round trip to every replica, assuming the leader is the last replica we try and all other followers are unsure of who the leader is.

**Comment by tbg** [comment]
@nvanbenschoten let's walk through this with some examples. n1, n2 are up to date, n3 is just coming back after a long hiatus.

If a request comes in at n3, and the group is not initialized yet, we will wake the leader [here](https://github.com/cockroachdb/cockroach/blob/455d9709d013eb55f5db3ce8f65ef252cc4ffd99/pkg/kv/kvserver/replica_send.go#L65) which means it will be discovered as soon as network latency allows (or, if no leader exists, an election will happen after the timeout). The lease request goes to the proposals map (and raft drops the proposal, since no leader known). Once the leader is known (which may be immediately since discovery was already initiated earlier), the lease either gets proposed or a NLHE comes back.

If I got everything right here, your proposal seems exactly right. I also don't think we're regressing from the status quo: the only wait is when the group as a whole doesn't have the leader and needs to time out for an election first. This is true after and before. This affects all of the ranges for which the node that went down was the leader, but also within the first 10 minutes of rebooting, it will unquiesce all of its raft groups due to the scanner, so after that requests wouldn't hit this random delay any more. (Still something to work on here, but since it's not regressing - no problem).

**Comment by nvanbenschoten** [comment]
That all checks out. So then you're ok with waiting synchronously for the leader election and then returning a more descriptive NLHE that points to the current leader?

Because if we are ok with that then this change actually becomes very targetted – it's just a small check somewhere near `handleRaftReady`. I think the only thing left to figure out would be how the synthetic NLHE would fit in with some of the speculative lease caching in the RangeCache that @andreimatei is working on.

**Comment by bobvawter** [comment]
Changing the `/health?ready=1` endpoint to report unready if the node is behind on snapshots would give operators an easy way to steer SQL traffic away from a node in this condition.

**Comment by drewdeally** [comment]
Is that in the latest release?

**Comment by andreimatei** [comment]
I think Bob was making a suggestion for something to implement; "it" does not exist.
I think the first problem with this suggestion is that it's unclear what it means for a node to be "behind on snapshots". There's legitimate situations when a node is supposed to be receiving a lot of snapshots but, at the same time, it should generally be serving traffic. Another problem is that it's entirely possible for a node to wedge a range by requesting a lease for it before finding out that "it is behind on snapshots", and then it doesn't matter whether further traffic is moved away from that node - the harm has already been done as the other replica are no longer able to serve traffic either.

I've been working on this (something similar to what had been discussed in previous comment around not attempting to acquire leases if a replica is not the leader), but hitting various subtleties. Still on it...
